### ISSUE TITLE
Co-located placements should have consistent shardstate

### ISSUE THREAD

**Issue by marcocitus** [citusdata/citus#810]
When we introduce table co-location, we need to ensure that the shardstate of co-located placements is always the same. Meaning that when we mark a placement as inactive, all co-located placements should also be marked inactive.

Consider we have two distributed, A and B. Each have 2 shards with 2 placements.

```
A1.1 A1.2 A2.1 A2.2
B1.1 B1.2 B2.1 B2.2
```

We can normally perform co-located joins between A and B. However, we are also independently writing to each table. After a while, 2 of the placements are marked as inactive due to temporal failures.

```
  x  A1.2 A2.1 A2.2
B1.1   x  B2.1 B2.2
```

At this point, joins between A1 and B1 that are normally handled by the router planner will be passed on to the logical planner where they are likely to error out with a feature "unsupported" error, or be executed using the real-time executor and return stale results. This depends on the query on whether the logical planner supports the query.

This situation will get a lot hairier when we perform other co-located operations such as INSERT/SELECT, shard rebalancing, and especially foreign keys. 

Consider that we may have independent inserts on A1 and B1 that violate the foreign key constraints, but this might not be detected because of the co-located placements being stale. Therefore this change is _required_ to make co-located foreign keys work with Citus.

**Comment by anarazel** [comment]
I don't think it is a good idea to mark all the colocated shards invalid. The lock escalation this will require will be painful, and the increased amount of shards having to be repaired will quite bad too.

If we are, and I'm not sure we immediately have to be, concerned about runtime errors we should rather add a check there making sure that colocated shards are healthy and error out if not. That should be fairly easy to add to router and I'd be surprised if it were different for INSERT ... SELECT.

**Comment by marcocitus** [comment]
I would much rather pay the price on failed writes (rare) then on every write (common).

**Comment by anarazel** [comment]
Which price? You mean the runtime cost of checking for failed shards?

How are you thinking of dealing with locking issues around marking all colocated shards as invalid? And about the increased recovery costs?

**Comment by marcocitus** [comment]
The point of marking all co-located placements as inactive is to keep one set of healthy co-located placements at all time. If we allow arbitrary shardstate changes as you're suggesting then all co-located operations such as writes to tables with foreign keys may eventually be unavailable. That is obviously unacceptable.

Instead of locking placement metadata for a particular shard, we should have a common lock for all co-located shards. Additionally, we should grab an exclusive lock prior to any write and then check whether the other set of placements is still healthy.

**Comment by anarazel** [comment]
> The point of marking all co-located placements as inactive is to keep
> one set of healthy co-located placements at all time.  If we allow
> arbitrary shardstate changes as you're suggesting then all co-located
> operations such as writes to tables with foreign keys may eventually
> be unavailable. That is obviously unacceptable.

I don't think that really changes the picture.

In your example of:
A1.1 A1.2 A2.1 A2.2
B1.1 B1.2 B2.1 B2.2

if you have one INSERT that succeeded to A1.1 but failed on A1.2, and
one one INSERT that succeeded on B1.2 but failed on B1.1, you're pretty
much screwed.

Either we allow partial colocations, or we accept invalid data, or we
only ever allow one concurrent write (even INSERT) to any shard of a any
colocated table.


(there are some independent issues in our code for the case where one
backends fails in A1.1 and the other in A1.2, but both succeed in the
other, but that's less likely, and more importantly independent)



> Instead of locking placement metadata for a particular shard, we
> should have a common lock for all co-located shards.

That'll be quite a contention point in the common case of all tables
being co-located, especially if we go for the "implicit colocation"
behaviour.


> Additionally, we
> should grab an exclusive lock prior to any write and then check
> whether the other set of placements is still healthy.

Doesn't this essentially boils down to not allowing *any* overlapping
writes to any individual shard?  To do this race-free you can't release
the lock during the DML, unless you use worker-side transactions. If you
do, you'll have to hold the lock over the COMMIT; sent to all involved
workers.  I mean that'll reduce our INSERT throughput by a magnitude. Or
am I missing something?

**Comment by marcocitus** [comment]
> if you have one INSERT that succeeded to A1.1 but failed on A1.2, and one one INSERT that succeeded on B1.2 but failed on B1.1, you're pretty much screwed.

No, after the insert on A1.2 failed both A1.2 and B1.2 are marked unhealthy. When the insert on B1.1 failed, there is no change to its shardstate, because it is the last healthy placement. 

> That'll be quite a contention point in the common case of all tables being co-located, especially if we go for the "implicit colocation" behaviour.

No, we need to grab a shard metadata lock anyway and this is always a share lock. It only conflicts with shard rebalancing. 

> Doesn't this essentially boils down to not allowing *any* overlapping writes to any individual shard?

No, this boils down to grabbing an exclusive lock after the write fails (rare) which should be the last thing you do before commit such that it's guaranteed to be released. Succeeding writes don't need to grab a lock (worst-case, they'll get redundantly sent to an unhealthy placement), nor do we have such a lock today. Serializing shardstate changes is something we should be doing regardless of co-location.

**Comment by anarazel** [comment]
On 2016-09-28 00:27:46 -0700, Marco Slot wrote:
> > if you have one INSERT that succeeded to A1.1 but failed on A1.2, and one one INSERT that succeeded on B1.2 but failed on B1.1, you're pretty much screwed.
>
> No, after the insert on A1.2 failed both A1.2 and B1.2 are marked unhealthy. When the insert on B1.1 failed, there is no change to its shardstate, because it is the last healthy placement.

Unless the other transaction has committed they can't see that the other
session has marked the placements as unhealthy.

**Comment by marcocitus** [comment]
> Unless the other transaction has committed they can't see that the other session has marked the placements as unhealthy.

Yeh, we probably need to use a new snapshot before checking the shardstate.

The issue isn't easy to solve, but it's critical for co-located operations to remain available under temporal failures (which unfortunately includes reaching max_connections on the worker) and foreign key constraints to be correctly honoured, so I guess we don't have much of a choice.

**Comment by sumedhpathak** [comment]
@marcocitus If I understand correctly, without fixing this, we can arrive at a situation where effectively the shards *aren't* co-located, and thus cannot guarantee FK constraints? 
What is your estimate for fixing this issue?

**Comment by marcocitus** [comment]
Correct. 1 week.

**Comment by marcocitus** [comment]
No longer relevant after #5381
